### ISSUE TITLE
Videos silently fail if the machine has no audio device

### ISSUE THREAD

**Issue by libTorrentUser** [renpy/renpy#5763]
I think this might be related to https://github.com/renpy/renpy/issues/2793

SDK 8.3.0 on Linux (Alpine Linux). If the machine does not have an audio device or it is not properly configured, videos will *SILENTLY* fail!

I just spent hours trying to figure out why videos where not being displayed. I was running the latest stable release of the SDK inside a QEMU virtual machine without an audio device and noticed that not even the tutorials videos where showing. I decided to manually try to play the video, just to see if at least the system was able to do it. Video players like `mpv` work in such machines without issues - they simply complain that there is no usable audio device and play the video anyway. And that was when it hit me - what if create a VM with audio?

And there you have it. The moment I added an audio device and configured it, renpy started playing the videos.

PS: simply adding the audio device is not enough. It must be properly configured too. On Linux, that means installing `alsa` and ensuring audio is working for the user. For me, that meant I needed to be able to successfully execute `alsamixer`. The moment I did, it all worked.

PS 2: is it just me or the sdk GUI needs `freeglut` to run? Without `freeglut`  (or some sort of `glut`), it segfaults. I didn't see it as a requirement and only noticed that was the cause when I installed the mesa demos to run the infamous `glxgears`, just to see if graphics were working correctly in the VM and that package had `freeglut` as a dependency.

**Comment by renpytom** [comment]
Glut isn't required, but GL is, so maybe that pulled something in? 

Audio is required - this is the sort of issue that you have to go out of your way (creating a vm) to cause.

**Comment by renpytom** [comment]
Reading through the SDL source code, it should fall back to the dummy driver if no audio driver is available. However, for it do do that, it needs to first detect that pulseaudio, alsa, and oss are all unavailable. My guess is one is present in your VM and not working, which is causing this problem. 

If you'd like, please do an 

lsof -n | grep renpy 

in the VM to see if there's an open audio device.

**Comment by libTorrentUser** [comment]
TLDR version
Not `glut` (or whatever variant) but `libxi` is what seems to be needed to avoid a segfault. At least on an Alpine Linux machine using DWM.

And you are absolutely right about the audio issue. Videos will work without audio devices, as long as no audio driver/package (alsa in my case) is installed.

Can you please investigate the segfault issue (could it also be related to SDL, only, this time, to the input subsystem???) and can we add some sort of note, somewhere, in a wiki or in a readme, regarding the video issue? A message box or perhaps add just another line to the logs saying that videos will not play because "Sound init" failed?
END OF TLDR


You are right, GLUT is not needed. What seems to be needed was  `libxi`. When I added the `freeglut` package, thinking it was the culprit, the only dependency it pulled was `libxi`. So, yeah, with `libxi` it works, without `libxi`, when I try to start `renpy` by itself or ``renpy pathToGame`, I can see a window opening for a microsecond and then segfaults.

By the way, there is no difference in the `log.txt` file when it crashes. The log shows exactly the same messages. Which is understandable, afterall, if the code had detected the cause, it would have logged it, and not crashed :)

As for the audio device issue.... AWESOME CATCH! If I remove the audio device from the qemu command line but install `alsa` anyway, videos will not work. If I do not install `alsa`, videos will work. I'm testing all this using  the `Tutorial` project and going to `Video playback`. When it fails and we take a look inside `log.txt`

```
Running _start took 0.00s
Sound init failed. Proceeding anyway.
Traceback (most recent call last):
  File "/home/unsafe/renpy-8.3.0-sdk/renpy/audio/audio.py", line 1015, in init
    renpysound.init(renpy.config.sound_sample_rate, 2, bufsize, False, renpy.config.equal_mono, renpy.config.linear_fades)
  File "renpysound.pyx", line 462, in renpy.audio.renpysound.init
  File "renpysound.pyx", line 111, in renpy.audio.renpysound.check_error
Exception: ALSA: Couldn't open audio device: No such file or directory
Interface start took 0.17s
```

Summing up
`no audio device + no alsa package`: OK
`no audio device + alsa package`: No videos

I'm not sure when I installed `alsa` in the VM I was using when I first opened this issue. I used the same command line I always used whenever I wanted an `Alpine Linux + DWM` VM, and it never had audio devices in it. Ever. So installing `alsa` would be the last thing I'd try there. But I guess something (or someone who might be one of the persons in this conversation that is not you) might have added the packages and I just didn't see. Very very strange. I tried to rollback the command line I was using before adding the audio device and I just couldn't reproduce the issue anymore. I mean, I could, if I installed `alsa`. But why it was there when I opened the issue will remain a mistery forever. Believe me, this is the very first time I added an audio device to a qemu VM. Ever!!!

Anyway..

Now, I might be wrong but, looking at the code https://github.com/renpy/renpy/blob/00750cb5d521bf705dba0f2f493a3aeac6ea4319/module/renpysound_core.c#L1322
and seeing how the functions were named, it seems pretty clear that playing videos is something that came later in the lifecycle and you guys just reused that internal sound API. Makes sense. It was already using ffmpeg and SDL anyway :) So that explains why video will not load even when the audio is muted (and the machine doesn't even have an audio device :)) - the internal API fails to initialize because it always tries to initialize SDL with audio.

Not trying to init the audio subsystem when the audio is "mute" or the error returned by SDL indicates that something went horribly wrong with the audio driver would, of course, be the best solution. But, since messing around with that part of the code could be very annoying (I could help if you wish), I propose a simpler solution that would make everybody happy(ier) - Show a message box (like those that appear when we use "quick save" or whatever) saying that videos will not play because the audio system could not be initialized. Some users might find it strange, but at least they will know what they have to fix in their system until, and if, renpy's core audio (and video) API is ever updated to handle such cases. 

Oh, and definitely add a new line to the logs saying the same thing - "NO VIDEO FOR YOU"

PS: honestly, I just realized the game I was playing had videos when I watched an youtube video and went "What? I never saw that cutscene! Is there a new version available???"

**Comment by renpytom** [comment]
Given that this is a problem that only occurs in misconfigured VMs, I'm probably not going to change Ren'Py to address it. I think the number of people playing games in VMs is small, and the number that are playing in VMs without working audio is smaller - so I don't think that it's worth complicating Ren'Py to try to detect this case.

**Comment by renpytom** [comment]
I've changed the logs to indicate that Audio and video init failed, not just sound. 

I'm not going to address the libXi problem - it's a standard part of X11, and if it's not there, your system is broken. It's also loaded from SDL2, so the problem would have to be addressed there. This may be fixed when I update to a newer version of SDL2 - we'll see. 

Fundamentally, Ren'Py assumes a standard Linux desktop, and that includes things like audio and libraries. Specifically, it's targeting Ubuntu and steam-runtime, so users of other distros should make sure similar libraries are available.

**Comment by libTorrentUser** [comment]
> I've changed the logs to indicate that Audio and video init failed, not just sound.

Nice! Would it be too hard to also show something more visible to regular users? I was thinking maybe showing something in one of those "message boxes" that appear when users do something like pressing the quick save button. Otherwise some users might never know their systems do not have everything Ren'Py requires work as the developers expected.

As for `libxi` you are right about it being port of X11. You are wrong when you think a system is broken if it is not there. In fact, it does not seem to be considered a required dependency on Alpine or Arch linux. In fact, it wasn't even supposed to be required by SDL2
https://github.com/libsdl-org/SDL/issues/8378
so, yeah, it seems that a SDL2 update is in order :)

Meanwhile, since it is kind of an obscure thing, and there is no such thing as a standard linux desktop, I just thought that, perhaps some sort of warning somewhere about it being required could help your users. If you want to limit Ren'Py to Ubunto and steam-runtime, that is, of course, your choice. But why do that when the only thing missing for the v8.3.0 SDK to work anywhere is a single lib that is not installed by default in some distros? 

I just think that you can save people a lot of headache by simply mentioning something like "segfaulting? Try installing libxi or wait until we update to a newer SDL2 version" somewhere. In a wiki maybe. 

Anyway, thanks for helping figuring out the problem. Technically, it is all SDL. Damn you, SDL!!! :)

**Comment by renpytom** [comment]
A popup is too much for the unlikely case of non-working audio. The log is fine. 

I'll look into taking that patch, but in general I have to set a standard for which Linux systems to support.

**Comment by libTorrentUser** [comment]
Ah, but the thing is, apparently you are already pretty much support them all. People on Arch laugh at how bloated Ubuntu is. Alpine people do the "Crocodile Dundee" on Arch people: "That is not lean? THIS is lean!"

And I just made it work on both. In fact, if I made it work on Alpine, I believe you'd be hard pressed to find one where it does not work. And the only thing causing problems wasn't really your fault. And the moment you patch the SDL, 100% portability :)

But back to the audio video issue, 
well, if you hated the pop-up, you will hate what I just came here to suggest... :) 
But I'll give it a try anyway

What about a warning screen just like the one that appears when you selected a renderer that the system cannot use? I removed GLES from the system, selected it as the rendering backend and got  a big warning screen saying 
```
Performance Warning
this computer is using software rendering
etc etc
```

One like that would be pretty awesome. The game will run as usual, but the user will certainly be aware that music and videos will not work. What I told you before is 100% true - I didn't even know the game I was testing had videos until I saw it on youtube.

**Comment by renpytom** [comment]
I'm not adding a warning outside of log.txt. 

I tried the libXi patch, but it doesn't apply to the version of SDL we're using now, so it will wait for an SDL upgrade. 

This is not a big issue. It happened once, on a vm, missing standard portions of X11. Ren'Py doesn't need to change because of problems with (virtual) hardware.
